internals-js/src/federation.ts

```diff
@@ -1736,6 +1737,100 @@ export class FederationBlueprint extends SchemaBlueprint {
       }
     }
 
+    const costFeature = schema.coreFeatures?.getByIdentity(costIdentity);
+    const costSpec = costFeature && COST_VERSIONS.find(costFeature.url.version);
+    const costDirective = costSpec?.costDirective(schema);
+    const listSizeDirective = costSpec?.listSizeDirective(schema);
+
+    // Validate @cost
+    for (const application of costDirective?.applications() ?? []) {
+      const parent = application.parent;
+      // @cost cannot be used on interfaces https://ibm.github.io/graphql-specs/cost-spec.html#sec-No-Cost-on-Interface-Fields
+      if (parent instanceof FieldDefinition && parent.parent instanceof InterfaceType) {
+        errorCollector.push(ERRORS.COST_APPLIED_TO_INTERFACE_FIELD.err(
+          `@cost cannot be applied to interface "${parent.coordinate}"`,
+          { nodes: sourceASTs(application, parent) }
+        ));
+      }
+    }
+
+    // Validate @listSize
+    for (const application of listSizeDirective?.applications() ?? []) {
+      const parent = application.parent;
+      // @listSize can only be applied to FIELD_DEFINITION
+      if (parent instanceof FieldDefinition && parent.type) {
+        const { assumedSize, slicingArguments = [], sizedFields = [] } = application.arguments();
+        // @listSize must be applied to a list https://ibm.github.io/graphql-specs/cost-spec.html#sec-Valid-List-Size-Target
+        if (!sizedFields.length && !isListType(parent.type)) {
+          errorCollector.push(ERRORS.LIST_SIZE_APPLIED_TO_NON_LIST.err(
+            `"${parent.coordinate}" is not a list`,
+            { nodes: sourceASTs(application, parent) },
+          ));
+        }
+
+        // Validate assumed size, but we differ from https://ibm.github.io/graphql-specs/cost-spec.html#sec-Valid-Assumed-Size.
+        // Assumed size is used as a backup for slicing arguments in the event they are both specified. 
+        // The spec aims to rule out cases when the assumed size will never be used because there is always
+        // a slicing argument. Two applications which are compliant with that validation rule can be merged
+        // into an application which is not compliant, thus we need to handle this case gracefully at runtime regardless.
+        // We omit this check to keep the validations to those that will otherwise cause runtime failures.
+        //
+        // With all that said, assumed size should not be negative.
+        if (assumedSize && assumedSize < 0) {
```

Comment:
Technically the logic of this is correct, but an `assumedSize` of 0 will not actually check it's size in the second part of the condition. Very much a nit, but I'd probably 
```suggestion
        if (assumedSize !== undefined && assumedSize < 0) {
```